File: Ejercicio3/3.cpp
#include <iostream>
#include <string>
using namespace std;

class Movil
{
public:
    string marca;
    string modelo;
    string sistemaoperativo;


    Movil(string _marca, string _modelo, string _sistemaoperativo) : marca(_marca), modelo(_modelo), sistemaoperativo(_sistemaoperativo) {}  // Constructor


    ~Movil() // Destructor
    {
        cout << "Se ha destruido un objeto de Movil: " << marca << " " << modelo << endl;  
    }

    void Detalles()
    {
        cout << "Marca: " << marca << endl;
        cout << "Modelo: " << modelo << endl;
        cout << "Sistema Operativo: " << sistemaoperativo << endl;
    }
};

class Telefono : public Movil
{
public:
    Telefono(string _marca, string _modelo, string _sistemaoperativo) : Movil(_marca, _modelo, _sistemaoperativo) {}  // Constructor

    ~Telefono() // Destructor
    {
        cout << "Se ha destruido un objeto de Telefono: " << marca << " " << modelo << endl;  
    }

    void Detalles()
    {
        Movil::Detalles();
    }

};

class Tablet : public Movil
{
public:
    Tablet(string _marca, string _modelo, string _sistemaoperativo) : Movil(_marca, _modelo, _sistemaoperativo) {}  // Constructor


    ~Tablet()  // Destructor
    {
        cout << "Se ha destruido un objeto de Tablet: " << marca << " " << modelo << endl; 
    }

    void Detalles()
    {
        Movil::Detalles();
    }

};

int main()
{
    cout << "--------Telefono-------" << endl;
    Telefono telefono01("Samsung", "Galaxy S21", "Android");
    telefono01.Detalles();

    cout << "--------Tablet-------" << endl;
    Tablet tablet01("Apple", "iPad Air", "iPadOS");
    tablet01.Detalles();

    return 0;
}

File: Ejercicio3/3.py
class Movil:
    def __init__(self, marca, modelo, sistemaoperativo):
        self.marca = marca
        self.modelo = modelo
        self.sistemaoperativo = sistemaoperativo

    def detalles(self):
        print("Marca:", self.marca)
        print("Modelo:", self.modelo)
        print("Sistema Operativo:", self.sistemaoperativo)

    def __del__(self):
        print("Se ha destruido un objeto de Movil:", self.marca, self.modelo)


class Telefono(Movil):
    def __init__(self, marca, modelo, sistemaoperativo):
        super().__init__(marca, modelo, sistemaoperativo)

    def detalles(self):
        super().detalles()

    def __del__(self):
        print("Se ha destruido un objeto de Telefono:", self.marca, self.modelo)


class Tablet(Movil):
    def __init__(self, marca, modelo, sistemaoperativo):
        super().__init__(marca, modelo, sistemaoperativo)

    def detalles(self):
        super().detalles()

    def __del__(self):
        print("Se ha destruido un objeto de Tablet:", self.marca, self.modelo)


print("--------Telefono-------")
telefono01 = Telefono("Samsung", "Galaxy S21", "Android")
telefono01.detalles()

print("--------Tablet-------")
tablet01 = Tablet("Apple", "iPad Air", "iPadOS")
tablet01.detalles()
